### PR TITLE
Add authorization tokens for API access

### PR DESCRIPTION
**Pull request by danlaine**
For documentation, see https://github.com/ava-labs/docs/pull/75

Addresses #176

### CODE REVIEW

File: api/auth/auth.go [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+	// ErrNoToken is returned by GetToken if no token is provided
+	ErrNoToken = errors.New("auth token not provided")
+
+	errWrongPassword = errors.New("incorrect password")
+)
+
+// Auth handles HTTP API authorization for this node
+type Auth struct {
+	lock     sync.RWMutex // Prevent race condition when accessing password
+	Enabled  bool         // True iff API calls need auth token
```
Comment (cgcardona) [inline api/auth/auth.go:38]: typo `iff`

Comment (danlaine) [reply api/auth/auth.go:38]: `iff` is `if and only if`, though admittedly this may be too jargony

Comment (cgcardona) [reply api/auth/auth.go:38]: Ah, I'll admit I didn't pick up on it. But glad I know it now.

**Comment by cgcardona** [comment]
* Generally not a best practice to show a password in plaintext
* Multiple passwords could potentially get confusing to some users
* Different password strength tests for the different passwords
  * Authorization token password
    * Not strength checked
  * User keystore passwords
    * Strength checked
* If a user calls `auth.changePassword` and then restarts their node then the password reverts to the original flag password and all the issued tokens under the new password are revoked

**Comment by danlaine** [comment]
@cgcardona 
`Generally not a best practice to show a password in plaintext`
Fixed in latest commit

`Multiple passwords could potentially get confusing to some users`
Yeah, I'm not really sure how to deal with this...I'm not sure how else to do this other than to have a password (or some other secret) known only by the node operator. I agree it could be confusing to have _another_ password protected thing besides the keystore.

`Different password strength tests for the different passwords`
Good point. Will add a commit for this, too.

`If a user calls auth.changePassword and then restarts their node then the password reverts to the original flag password and all the issued tokens under the new password are revoked`
I think this is OK for now but not ideal

**Comment by danlaine** [comment]
@cgcardona Added password strength checking

**Review by StephenButtolph — APPROVED** [review]

**Pull request merged**